pkg/fs/real.go

```diff
@@ -78,6 +80,24 @@ func (realFilesystem) Walk(root string, walkFn filepath.WalkFunc) error {
 	return filepath.Walk(root, walkFn)
 }
 
+func (realFilesystem) Lock(path string) (r Releaser, existed bool, err error) {
+	r, existed, err = flock.New(path)
+	r = deletingReleaser{path, r}
+	return r, existed, err
+}
+
+type deletingReleaser struct {
+	path string
+	r    Releaser
+}
+
+func (dr deletingReleaser) Release() error {
+	if err := dr.r.Release(); err != nil {
+		return err
+	}
+	return os.Remove(dr.path)
```

Comment:
Maybe I don't have things straight but this still feels racy. Should the flock package remove this while holding the lock, like around [here](https://github.com/prometheus/prometheus/blob/7eb849e6a82094b649087e36d2ce5132fc798ef6/util/flock/flock_unix.go#L31)? Surprised there's no *os.File.Remove to reference here.